libcontainer/cgroups/utils.go

```diff
@@ -118,6 +118,93 @@ func FindCgroupMountpointDir() (string, error) {
 	return "", NewNotFoundError("cgroup")
 }
 
+func FindCgroup2MountpointDir() (*MountLine, error) {
+	f, err := os.Open("/proc/self/mountinfo")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		text := scanner.Text()
+		parsedLine, err := parseMountLine(text)
+		if err != nil {
+			return nil, err
+		}
+		if parsedLine.FileSystemType == "cgroup2" {
+			return &parsedLine, nil
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return nil, NewNotFoundError("cgroup2")
+}
+
+type MountLine struct {
+	MntId       int
+	MntParentId int
+	MajorStDev  int
+	MinorStDev  int
+	Root        string
+	MountPoint  string
+	// options are typically a comma seperated list, optionally as key-value, or single arguments
+	// let the user parse them
+	MountOptions   string
+	OptionalFields map[string]string
+	FileSystemType string
+	MountSource    string
+	SuperOptions   string
+}
+
+func indexOf(item string, slice []string) int {
+	for idx, val := range slice {
+		if val == item {
+			return idx
+		}
+	}
+	return -1
+}
+
+func parseMountLine(line string) (MountLine, error) {
```

Comment:
Just a side thought: `/proc/self/mountinfo` is parsed at several places already:
- `FindCgroupMountpointDir()`
- `FindCgroupMountpointAndRoot()`
- `parseMountTable()`

rkt has also [yet another implementation](https://github.com/rkt/rkt/blob/v1.29.0/pkg/mountinfo/mountinfo.go). It could be good to factorize at least within runc.

If `MountPoint` has spaces, new lines or other special characters, `mountinfo` escapes them but the parser will not unescape them at the moment. If it is factorized, the unescaping could be fixed in one place.